### PR TITLE
[KYUUBI #5088][FLINK] Add result fetch timeout for Flink engine

### PR DESCRIPTION
**Pull request by link3280**
### _Why are the changes needed?_
As titled.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by link3280** [comment]
cc @pan3793 @bowenliang123

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/result/ResultSetUtil.scala [hunk 1]
```diff
@@ -84,34 +89,10 @@ object ResultSetUtil {
         case _: InterruptedException => fetched.getNextToken == null
       }
     } while (
-      fetched.getNextToken != null &&
+      System.currentTimeMillis() - fetchStart < fetchTimeout &&
```
Comment (cxzl25) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/result/ResultSetUtil.scala:92]: Will an exception be thrown if it times out, or will the state of the operation be the timeout state?

Comment (link3280) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/result/ResultSetUtil.scala:92]: Nope, then Flink engine would stop fetching more rows and return the result set.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -1319,6 +1319,14 @@ object KyuubiConf {
       .intConf
       .createWithDefault(1000000)
 
+  val ENGINE_FLINK_FETCH_TIMEOUT: ConfigEntry[Long] =
+    buildConf("kyuubi.session.engine.flink.fetch.timeout")
+      .doc("Result fetch timeout. Flink would keep fetching result until " +
+        "either the max rows limit is reached or the timeout is reached.")
+      .version("1.8.0")
+      .timeConf
+      .createWithDefault(Duration.ofMinutes(5).toMillis)
```
[the remark this answers is not in the document]
Comment (link3280) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1328]: It looks like we need to make a trade-off between batch queries and stream queries. For batch queries (bounded), the result should be complete, and the timeout should be optional or set to a greater value. While for stream queries (usually unbonded), the result would never be complete, thus select queries are just used for taking a peek and the completeness is not a problem.

[the remark this answers is not in the document]
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1328]: I think it's not a problem if we implement the incremental fetch mechanism, the client displays the reach records immediately and waits for the next batch until the server return EOF, user could cancel the stream query at any time if they don't want more results.

[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1328]: > this actually could cut the result off and silently returns in-completed results to the client. can we make it optional and leave the default value as None?

I agree with @pan3793 . Undefined as default value provides more flexibility.

**Comment by bobo495** [comment]
Looking forward to this new feature, which can be useful for querying streaming data！

**Comment by link3280** [comment]
Closed in favor of #4806.

**Pull request closed without merging**